torch/optim/functional.py

```diff
@@ -96,3 +96,39 @@ def adam(params: List[Tensor],
         step_size = lr / bias_correction1
 
         param.addcdiv_(exp_avg, denom, value=-step_size)
+
+
+def sgd(params: List[Tensor],
+        d_p_list: List[Tensor],
+        momentum_buffer_list: List[Optional[Tensor]],
+        weight_decay: float,
+        momentum: float,
+        lr: float,
+        dampening: float,
+        nesterov: bool):
+    r"""Functional API that performs SGD algorithm computation.
+
+    See :class:`~torch.optim.SGD` for details.
+    """
+
+    for i, param in enumerate(params):
+
+        d_p = d_p_list[i]
+        if weight_decay != 0:
+            d_p = d_p.add(param, alpha=weight_decay)
+
+        if momentum != 0:
+            buf = momentum_buffer_list[i]
+
+            if buf is None:
+                buf = torch.clone(d_p).detach()
```

Comment:
nit: why the detach here? Are there any case where this function is called outside of a `torch.no_grad()` block?

Reply:
not really, it should all under a `torch.no_grad()` block, I just copied this from the sgd.py, if there's no need to detach here, I can remove it.

Reply:
Ho I didn't realize it was in the original version. I think you should keep it as is here to reduce any chance to break stuff in this refactor. :)